firebase-firestore/src/main/java/com/google/firebase/firestore/model/MutableDocument.java

```diff
@@ -73,6 +78,46 @@ private MutableDocument(
     this.value = value;
   }
 
+  /**
+   * Creates a document with no known version or data, but which can serve as base document for
+   * mutations.
+   */
+  public static MutableDocument newInvalidDocument(DocumentKey documentKey) {
+    return new MutableDocument(
```

Comment:
These factory methods appear to duplicate the logic defined in their corresponding `convertToXXX()` methods. To be a bit more "dry" (go/dry), consider reviving the constructor that takes the `DocumentKey`, changing its visibility to `private`, using that constructor, and then calling the corresponding `convertToXXX()` method, returning its result.

Reply:
I opted to not call into the one argument constructor since keeping it initializes some of the values twice. The one argument constructor would call the default constructor, which then creates new objects, only to be immediately destroyed by the mutating function.

Furthermore, as least some of the code duplication would have to be moved to the one argument constructor.

Reply:
How about this idea. The only constructor is `MutableDocument(DocumentKey)`, which initializes `this.key` and does not assign anything to the other instance variables (they will be initialized to null). If this constructor is called from the factory methods, followed by the corresponding `convertToXXX()` method, then all instance variables will be initialized and none will be multiply initialized. I realize that this leaves the object temporarily in a weird state after calling the constructor; however, since it is private you have complete control over how the newly-constructed object is used and can prevent ever exposing an object in this intermediate state.

Reply:
I like that. Updated.